File: programmers/etc/level4/트리_트리오_중간값.py
from collections import deque

def bfs(tree, start):
    distance = []
    q = deque()
    q.append([start, 0])
    visited = {start: True}    
    while q:
        cur, dist = q.popleft()
        distance.append([cur, dist])        
        for nxt in tree[cur]:
            if nxt not in visited:
                q.append([nxt, dist+1])
                visited[nxt] = True
    return distance

def solution(n, edges):
    tree = {i+1: [] for i in range(n)}
    for edge in edges:
        tree[edge[0]].append(edge[1])
        tree[edge[1]].append(edge[0])
    
    d1 = bfs(tree, 1)
    d2 = bfs(tree, d1[-1][0])
    
    if d1[-1][1] == d1[-2][1]:
        return d2[-1][1]
    else:
        return d2[-2][1]
    

n, edges = 4, [[1,2],[2,3],[3,4]]
print(solution(n, edges)) # 2

n, edges = 5, [[1,5],[2,5],[3,5],[4,5]]
print(solution(n, edges)) # 2

File: Java/programmers/etc/level4/트리트리오중간값.java
package CodeTest.Java.programmers.etc.level4;

import java.util.*;

public class 트리트리오중간값 {
    private static Map<Integer, List<Integer>> tree;
    private static List<int[]> distance;
    private static Queue<int[]> q;
    private static Map<Integer, Boolean> visited;

    public int solution(int n, int[][] edges) {
        tree = new HashMap<>();
        for (int i=0; i<n; i++) {
            tree.put(i+1, new ArrayList<>());
        }

        for (int[] edge : edges) {
            List<Integer> u = tree.get(edge[0]);
            u.add(edge[1]);
            tree.put(edge[0], u);

            List<Integer> v = tree.get(edge[1]);
            v.add(edge[0]);
            tree.put(edge[1], v);
        }

        List<int[]> d1 = bfs(1);
        List<int[]> d2 = bfs(d1.get(d1.size()-1)[0]);

        boolean flag = d1.get(d1.size()-1)[1] == d1.get(d1.size()-2)[1];

        return flag ? d2.get(d2.size()-1)[1] : d2.get(d2.size()-2)[1];
    }

    private static List<int[]> bfs(int start) {
        distance = new ArrayList<>();

        q = new LinkedList<>();
        q.add(new int[]{start, 0});
        
        visited = new HashMap<>();
        visited.put(start, true);

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            distance.add(cur);

            int cur_node = cur[0], cur_dist = cur[1];
            for (Integer nxt_node : tree.get(cur_node)) {
                if (!visited.containsKey(nxt_node)) {
                    q.add(new int[]{nxt_node, cur_dist+1});
                    visited.put(nxt_node, true);
                }
            }
        }

        return distance;
    }
}
